daemon/api_quota_test.go

```diff
@@ -0,0 +1,239 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/snapcore/snapd/daemon"
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/overlord/servicestate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap/quota"
+	"gopkg.in/check.v1"
+)
+
+var _ = check.Suite(&apiQuotaSuite{})
+
+type apiQuotaSuite struct {
+	apiBaseSuite
+}
+
+func (s *apiQuotaSuite) SetUpTest(c *check.C) {
+	s.apiBaseSuite.SetUpTest(c)
+	s.daemon(c)
+}
+
+func mockQuotas(st *state.State, c *check.C) {
+	grp0, err := quota.NewGroup("foo", 9000)
+	c.Assert(err, check.IsNil)
+	grp1, err := grp0.NewSubGroup("bar", 1000)
+	c.Assert(err, check.IsNil)
+	grp2, err := quota.NewGroup("baz", 2000)
+	c.Assert(err, check.IsNil)
+	c.Assert(servicestate.UpdateQuotas(st, grp0, grp1, grp2), check.IsNil)
+}
+
+func (s *apiQuotaSuite) TestPostQuotaUnknownAction(c *check.C) {
+	data, err := json.Marshal(daemon.PostQuotaGroupData{Action: "foo", GroupName: "bar"})
+	c.Assert(err, check.IsNil)
+
+	req, err := http.NewRequest("POST", "/v2/quotas", bytes.NewBuffer(data))
+	c.Assert(err, check.IsNil)
+	rsp := s.errorReq(c, req, nil)
+	c.Assert(rsp.Status, check.Equals, 400)
+	c.Check(rsp.Result.(*daemon.ErrorResult).Message, check.Equals, `unknown action "foo"`)
+}
+
+func (s *apiQuotaSuite) TestPostQuotaInvalidGroupName(c *check.C) {
+	data, err := json.Marshal(daemon.PostQuotaGroupData{Action: "ensure", GroupName: "$$$"})
+	c.Assert(err, check.IsNil)
+
+	req, err := http.NewRequest("POST", "/v2/quotas", bytes.NewBuffer(data))
+	c.Assert(err, check.IsNil)
+	rsp := s.errorReq(c, req, nil)
+	c.Assert(rsp.Status, check.Equals, 400)
+	c.Check(rsp.Result.(*daemon.ErrorResult).Message, check.Matches, `invalid quota group name: .*`)
+}
+
+func (s *apiQuotaSuite) TestPostEnsureQuotaUnhappy(c *check.C) {
+	daemon.MockServicestateCreateQuota(func(name string, parentName string, snaps []string, memoryLimit quantity.Size) error {
+		c.Check(name, check.Equals, "booze")
+		c.Check(parentName, check.Equals, "foo")
+		c.Check(snaps, check.DeepEquals, []string{"bar"})
+		c.Check(memoryLimit, check.DeepEquals, quantity.Size(1000))
+		return fmt.Errorf("boom")
+	})
+
+	data, err := json.Marshal(daemon.PostQuotaGroupData{
+		Action:    "ensure",
+		GroupName: "booze",
+		Parent:    "foo",
+		Snaps:     []string{"bar"},
+		MaxMemory: 1000,
+	})
+	c.Assert(err, check.IsNil)
+
+	req, err := http.NewRequest("POST", "/v2/quotas", bytes.NewBuffer(data))
+	c.Assert(err, check.IsNil)
+	rsp := s.errorReq(c, req, nil)
+	c.Check(rsp.Status, check.Equals, 400)
+	c.Check(rsp.Result.(*daemon.ErrorResult).Message, check.Matches, `boom`)
+}
+
+func (s *apiQuotaSuite) TestPostEnsureQuotaHappy(c *check.C) {
+	var called bool
```

Comment:
(nitpick) maybe use a int here so that we can check that it wasn't called multiple times